backend/app/service/upgrade.go

```diff
@@ -205,6 +207,8 @@ func (u *UpgradeService) handleRollback(originalDir string, errStep int) {
 	if err := common.CopyFile(path.Join(originalDir, "1pctl"), "/usr/local/bin"); err != nil {
 		global.LOG.Errorf("rollback 1panel failed, err: %v", err)
 	}
+	_, _ = cmd.Execf("cp -r %s /usr/local/bin", path.Join(originalDir, "lang"))
+
 	if errStep == 2 {
 		return
 	}
```

Comment:
The code appears to be free of any immediate irregularities, issues, or optimizations but may need minor modifications such as consistent use in the `fileOp` parameter.

The code seems well-established with no apparent errors; it handles rolling back upgrades and restoring a backup using common.CopyFile. However, there is one change recommended that does not seem critical:

- Instead of using double underscores (`_`) when calling `.Execf()` methods directly within a function's body which would cause potential syntax violations, you can pass the `cmd`, arguments as strings instead. For example: 

`if _, err := cmd.Execf(strings.Replace(cmd.Path(), " ","\\ ", -1), fmt.Sprintf("%s %s", args...)); err != nil {`

This way, even though `_` characters will still exist after parsing the variable reference, they have been effectively replaced before being assigned their actual value from `cmd.Path()`. This should solve problems where these are expected to clash with variables already defined elsewhere in the same scope (like functions).

Additionally, consider reviewing other parts of the code for better readability and efficiency, if necessary. In the current context, this code doesn't appear to significantly deviate from typical practices.